### PR TITLE
make changes to allow optimization for both CPU and Memory configurations for GCP Cloud Functions V2

### PR DESCRIPTION
**Pull request by hdu3**
# Changes
* Create new CPU-Memory-Duration function to model the serverless function's execution time based on the CPU and Memory allocation.
* Create new classes to support 2D search for the CPU-Memory space.
* Use Google Cloud `functions_v2` API to programmatically update both the CPU and Memory configuration for the function being optimized.
* Add new flag `GCPv2` to enable 2D optimization.
* Add unit tests for the new Classes.

### CODE REVIEW

**Review by arshiamoghimi — CHANGES_REQUESTED** [review]
The code looks good. Thanks. Other than the minor comments inlined, here are some items you need to fix:
- add documentations in the READMEs (main and configuration at src/configuration/README.md).
- pull the latest version of the main branch
- Fix the merge conflicts

File: src/configuration/defaults.py [hunk 1]
```diff
@@ -5,8 +5,8 @@
     "coefficient_of_variation_threshold": 0.05,
 }
 MAX_NUMBER_OF_INVOCATION_ATTEMPTS = 5
-MAX_TOTAL_SAMPLE_COUNT = 20
-MIN_SAMPLE_PER_CONFIG = 4
-TERMINATION_THRESHOLD = 3
+MAX_TOTAL_SAMPLE_COUNT = 250
+MIN_SAMPLE_PER_CONFIG = 5
+TERMINATION_THRESHOLD = 4
```
[the remark this answers is not in the document]
Comment (hdu3) [reply src/configuration/defaults.py]: Thanks for the call out. Changed.

File: src/exploration/explorer.py [hunk 2]
```diff
@@ -31,6 +31,7 @@ def __init__(
         self.payload = payload
 
         self.memory_space = np.array(list(memory_space), dtype=int)
+        # self.cpu_memory_space = np.array()
```
[the remark this answers is not in the document]
Comment (hdu3) [reply src/exploration/explorer.py]: Removed.

File: src/exploration/explorer_2d.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                try:
+                    results.append(future.result())
+
+                except InvocationError as e:
+                    logger.debug(e)
+                    if error is None:
+                        error = e
+                    # self.cost += self.price_calculator.calculate_price(
+                    #     self._memory_config_mb, e.duration_ms, self._cpu_config
+                    # )
```
[the remark this answers is not in the document]
Comment (hdu3) [reply src/exploration/explorer_2d.py]: Removed.

File: src/exploration/explorer_2d.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        # If one thread raises an invocation error we raise it.
+        if error:
+            raise error
+
+        # Calculate the cost
+        # self.cost += np.sum(
+        #     self.price_calculator.calculate_price(
+        #         self._memory_config_mb, np.array(results), self._cpu_config
+        #     )
+        # )
```
[the remark this answers is not in the document]
Comment (hdu3) [reply src/exploration/explorer_2d.py]: Removed.

**Comment by hdu3** [comment]
> The code looks good. Thanks. Other than the minor comments inlined, here are some items you need to fix:
> 
> * add documentations in the READMEs (main and configuration at src/configuration/README.md).
> * pull the latest version of the main branch
> * Fix the merge conflicts

Thanks for the comment. I've addressed these comments in the new commit.

**Review by arshiamoghimi — APPROVED** [review]

**Pull request merged**